### ISSUE TITLE
[Proposal] Safe Lanes

### ISSUE THREAD

**Issue by bobbens** [naev/naev#897]
(taken from the [old wiki](https://web.archive.org/web/20130218201416/http://wiki.naev.org/wiki/Safe_Lanes))
This is a proposal for "safe", patrolled flight lanes within systems. It relies on and reinforces the proposal about improved targeting and navigation, and the detailed article about asteroids.

## Goals
* Resolve jump point combat issues
* Allow weaker ships to survive more reliably in faction protected systems
* Generate opportunities for contraband and piracy related activities
## Safe Lanes
A safe lane is a route within a system that is heavily patrolled by a law-enforcing organization (usually a faction). Ships traveling along these lanes will be protected from pirate attack, because there will always be friendly combat ships nearby. In addition, safe lanes are marked by turret emplacements at set intervals, so if a pirate force does attempt a raid on ships traveling along the lane, those ships can try to find shelter around the closest turret emplacement.

Safe lanes connect places of interest, notably jump points and planets, and will exist in any system that has a sufficient lawful presence. However, a place of interest needs not necessarily have a direct safe lane to another one, so it may be necessary to detour through the system to reach the destination via safe lanes.

Positive effects of safe lanes include:

* Weaker ships can travel more outlying systems in relative safety
* There is an opportunity for generated system patrol missions involving safe lanes
* There is an opportunity for generated pirate activities involving safe lanes

Reasons why the player might NOT make use of safe lanes:

* Safe lanes may not interconnect some places of interest, such as jump points leading to non-faction systems
* Smuggling contraband through safe lanes is doomed to failure
* Time-critical missions may require a direct route to a destination, while safe lanes would typically be suboptimal
* The player would likely prefer dealing with pirates off the beaten track than with lanes swarming with enemy faction ships

## Building Safe Lanes
There are two approaches to building safe lanes:

### Manually
Lanes would be manually "painted" in the system, using the ingame editor. The advantage of this is that lanes can be built exactly as intended for that system, and can be adjusted as required. The downside is that it implies manual maintenance.

## Automatically
An automatic generation of safe lanes constitutes a significant reduction in maintenance, but the results may not be desirable. Any automatic generation of safe lanes should try to satisfy the following:

* Each safe lane connects one place of interest with one other. Places of interest include:
   * Jump points
   * Planets or stations owned by the faction patrolling the safe lane
   * Planets or stations owned by factions friendly to the faction patrolling the safe lane
   * Hypergates
* A safe lane must have turret emplacements at regular distances.
* A safe lane may need bends in order to path around obstacles such as asteroid fields.
* The maximum number of safe lanes in a system should scale with faction presence, with diminishing returns.
* Safe lanes should prioritize interconnecting planets with jump points leading to friendly systems.
* Each safe lane must connect to at least one asset owned by the faction patrolling the safe lane (e.g. Empire doesn't patrol between a Dvaered station and a jump point).

## Pirate Spawning
The primary purpose of safe lanes being a resolution to the rampant pirate issues, pirate spawning should be revised. Rather than letting pirates jump into a system like any other ships, they should be spawned directly in a system, away from the safe areas, coming out of nowhere. This seems unreasonable, and it is, but the current problems with jump point combat justify it. Additionally, a safe lane to a jump point would likely make it impossible for pirates to enter through the jump point and survive, which is undesirable.

A second consideration is that piracy should be more widespread. While currently the various factions have large, pirate free cores, safe lanes remove the need for these. After all, the core systems have a very high faction presence, and as such will have many safe lanes. Trading within these systems will remain as safe as it is now, while straying away from the safe lanes will become dangerous. This is important, because smuggling contraband would be too easy in core systems otherwise.

**Comment by BTAxis** [comment]
Nice, well written proposal. Sound like a great idea, too!

**Comment by ids1024** [comment]
@nenau had a partly working implementation: https://github.com/nenau/naev/tree/safe_lanes

(I would like to the relevant forum thread if I could).

That was limited by a somewhat slow pathfinding algorithm, and adding more objects to the game. And it lacked any manual mechanism.

> An automatic generation of safe lanes constitutes a significant reduction in maintenance, but the results may not be desirable. Any automatic generation of safe lanes should try to satisfy the following:

At the very least, I think we want to manually specify which systems are connected by safe lanes (then it could automatically choose which assets should be, and determine a good route for the lanes to take).

**Comment by nenau-again** [comment]
Did I read "Safe Lanes" ?
(I got a problem with my account "nenau", so I had to create a second one)

I remember I used to code that pathfinder stuff and an algo that generates the best network to minimize average path for the traffic.
As it used to be my first "deep" contribution to the engine, I would not at all guarantee my implementation was totally free of memory leaks however...
Anyway, I had also implemented all the stuff from pirate and patrols behaviour to autonav with the lanes.
And yes, the automatic generation was very expensive in terms of CPU (you have to run the pathfinder n(n+1)/2 times at each minimization step, and there may be ~100 steps, n being the nb of assets). On my old laptop, it used to take ~3 seconds in systems with loads of assets.

The idea I got was to call the lanes generation algo on the system editor (with a special button), and store the lanes as data on the system file. This approach only poses problem for assets generated by diffs...

Second question is about the asteroids fields we have now (btw, did you add some outside Zied ?). My idea was to use an Uzawa algo to solve the inequality constraint problem that gives the shape of a lane that has to go around a polygon (that represent the asteroid field. That is of course a very funny problem, but I'm not sure I could have enough patience to code such a thing (do we have support for linear algebra ?)

So in short :
* The work I had done could be used in some sense to implement safe lanes, but it should be re-worked
* The asteroid technical problem can be solved
* But I'm not even sure I will gather enough free time to hard code like in the past again...

**Comment by PhoenixRiver** [comment]
> * Each safe lane connects one place of interest with one other. Places of
> interest include:
>    * Jump points

I should note that from memory there are a few jump points that are
meant to be "secret" - and difficult to find... they should not have
safe lanes running straight to them; maybe the current variable
affecting how easy they are to find should also affect the chances of
getting a safe lane?

I'll also agree that these safe lanes should be pregenerated or
generated only once and then written to file; it could be done on the
fly the first time you enter a system (with maybe a message "Scanning
new system for points of interest" or some such other distraction for
the player to cover the extra couple of seconds it might take - and
also as a debug for devs testing changes to a system) or else on load
(though this scales poorly - the more systems Naev develops, the
longer it takes) or else maybe it should be a file included in the git
repo? (risky as it may well become out of date more easily if not
maintained).

Just my tu'penny,

Phoenix

**Comment by nenau-again** [comment]
Ok, I have made a py script that generates lanes for the whole universe at once. It means that the whole lanes mesh is coherent, with trade routes between populated systems being more guarded than routes between empty systems. Here an example of what it gives on Goddard : 

![lanes_goddard](https://user-images.githubusercontent.com/60405677/78637104-044ab600-78aa-11ea-814b-07f4857172fc.png)

One sees the 3 jumps out of the system being protected, as well as the inhabited assets, in a coherent mesh.

Here is an other result in a system I have barely heard of (it's in Zlk space, I think?)

![lanes_armorhead](https://user-images.githubusercontent.com/60405677/78637481-b8e4d780-78aa-11ea-9f89-7924c48f3d5b.png)

One sees 3 jumps and 2 assets protected by a coherent lanes mesh again.

The script still has many issues to solve, and loads of work is still needed, but it could ultimately be used to generate lanes in the xml. Only problem are the unidiffs...

What do you think ?

**Comment by BTAxis** [comment]
That looks pretty promising. Is it capable of creating redundant routes within a system, or conversely, generating only a partial network for a system? As I recall the original proposal called for these things based on faction presence, to represent a faction's capability (or lack thereof) to nominally police a system.

**Comment by nenau-again** [comment]
It depends on the "money" I give to the lane generator. For example, here is what Goddard would do if they had funding issues :

![lanes_goddard_poor](https://user-images.githubusercontent.com/60405677/78649531-dbccb700-78bd-11ea-82ca-6d89e8707b97.png)

And if they're richer : 
 
![lanes_goddard_rich](https://user-images.githubusercontent.com/60405677/78649561-e8510f80-78bd-11ea-8bfe-04a23d813338.png)

Right now, I give manually the "money", but ultimately, it will be linked to presence. And what's more, I plan to have each faction building its lanes.

**Comment by BTAxis** [comment]
Very nice.

> And what's more, I plan to have each faction building its lanes.

What does this mean?

**Comment by nenau-again** [comment]
> What does this mean?
Right now, lanes don't belong to a particular faction, but that will be changed, which will slightly increase the complexity of the code, but not its principle. So for example, Empire, Goddard and Dvaer lanes would coexist in this system.

**Comment by nenau-again** [comment]
Just for fun : if I remove the lanes that make it possible to join Dvaered core and Frontier from Ogat (that is on the other side of the jump southeast of Goddard) we get that : 

![lanes_goddard_noogat](https://user-images.githubusercontent.com/60405677/78654416-c313cf80-78c4-11ea-8d89-d8f94ec1dc61.png)

As a matter of fact, the algorithm has lost most of its interest in this jump point to Ogat and prefers to secure directly the road from Arcturus to Delta Pavonis (West)

**Comment by ids1024** [comment]
Just looking at those examples, this looks like a nice design!

> I have made a py script that generates lanes for the whole universe at once. It means that the whole lanes mesh is coherent, with trade routes between populated systems being more guarded than routes between empty systems.

This sounds like it would work better than the earlier design.

We also have to consider how much manual control we would want over safe lane generation.

We probably want to be able to guarantee certain planets end up as part of the safe lane. Tuning attributes of the asset (presence, etc.) to make its significance clear may be sufficient.

With the earlier safe lane implementation, I wondered if it might be better to manually specify which systems should be connected by safe lanes, so the overall network seems well planned. Again, it will depend on what the generated version looks like, and could be addressed by tuning parameters of the system that effect generation.

> The script still has many issues to solve, and loads of work is still needed, but it could ultimately be used to generate lanes in the xml.

It's worth thinking about doing it in C at game load time, when jumping into the system, etc. If it's feasible. But doing it at build time does allow a more elaborate (i.e. time consuming) approach. So it's likely a good approach, but it's worth thinking about.

Relatedly: how is this going to affect the experience of someone using the universe editor to modify space, when they aren't really a programmer?

> Only problem are the unidiffs...

If we have relatively few unidiffs, it should be fine just to generate a separate lane network for the system with and without the diff, and only have separate copies if the result is different. I think this works fine with how we expect the unidiff system to be used, for the time being.

> So for example, Empire, Goddard and Dvaer lanes would coexist in this system.

Would these connect? How? This could be interesting (probably in a good way) at a point in the game when factions are at war (though they may also destroy each other's lanes...)

**Comment by BTAxis** [comment]
> Relatedly: how is this going to affect the experience of someone using the universe editor to modify space, when they aren't really a programmer?

I think that's a good point. If possible, I'd like to avoid requiring a python script call every time the galaxy is edited. Ideally we'd have a C implementation that's fast enough to work during gameplay, (which might be doable since we only need to solve for one system at a time), or failing that, a C implementation that runs whenever the galaxy is saved in the editor.

**Comment by ids1024** [comment]
If the implementation is fairly fast, and in C, you also see the safe lanes change dynamically with each adjustment in the editor. Which would be quite convenient, even if you're also familiar with Python and comfortable executing the script.

Of course, that means it needs to be a fairly fast implementation, and we can't use the nice numerical computation libraries available in Python.

For comparison, ship polygon generation is per-computed using Python now, but that doesn't inherently make anything harder (at least if we integrate it better) since generating sprites requires executing Python code anyway.

**Comment by nenau-again** [comment]
> Of course, that means it needs to be a fairly fast implementation, and we can't use the nice numerical computation libraries available in Python.

Yep, well... here comes the funny part.
Unlike previous design, this algo computes lanes for the whole galaxy at once, and in order to keep the run time short (and also for other reasons), I had to abandon the Djikstra pathfinding algo for something that consists in the resolution of a linear impedance problem (like when we solve a problem with resistances in electronics).
And for that, I need a sparse linear algebra library (I'm not cowboyish enough to code one from scratch). Here, I use scipy, which I'm quite familiar with, but I'm sure there exist some in C, it will simply be more complicated to find the right one...

About the question of run time, it's about a few seconds to generate all lanes in the universe, so it's doable at loading, but not really in real time... (of course, this can be changed if we switch to C)

And the algo itself uses a few not-so-trivial math. I will write a report about it if you're interested, but I'm not sure it will be readable by many people...

**Comment by nenau-again** [comment]
> Would these connect? How? This could be interesting (probably in a good way) at a point in the game when factions are at war (though they may also destroy each other's lanes...)

One can add rules at will (for example a faction is not allowed to build a lane that leads to an other faction's asset).
If there are lane-building factions at war in a system, the idea I had was to simply not build any lane at all because the system would basically be a battlefield. If one faction is in vast minority, the other faction builds lanes (for example in Fortitude, the Empire would build lanes, in C-whatsoever systems, the collective would build, and in Hades, no one builds anything). That would make systems at war more dangerous for traders.

**Comment by BTAxis** [comment]
> About the question of run time, it's about a few seconds to generate all lanes in the universe, so it's doable at loading, but not really in real time... (of course, this can be changed if we switch to C)

My thinking is that it might be best done when loading a save, and when applying a unidiff. It seems impractical to maintain "safe-lane-diffs", that could get messy fast what with various permutations and such. Better to just generate it on demand. A few seconds of lag on load and when applying a unidiff is acceptable. If it turns out sufficiently fast, we can think about more dynamic changes.

**Comment by bobbens** [comment]
If it is linear algebra-based (likely given it's solving an impedence problem), something like lapack (https://www.netlib.org/lapack/) would be a fairly standard option that should work well. However, it is likely that the matrices will be sparse due to being graph-based, in that case, what about csparse (https://people.sc.fsu.edu/~jburkardt/c_src/csparse/csparse.html)? Funny thing is that we actually have csparse included in naev and can also link to it, so it wouldn't require changing any dependencies. I added it back in the day to try to get an impedence-based economy working, but was never too happy with it.

**Comment by nenau-again** [comment]
Mm I'm so unfamiliar with compiled code that I didn't know it was possible to use a Fortran library like Lapack in C. but anyway, if we already have csparse included, then the best is to work with it. I'll look at that when I have time.
I guess I'll start with a function that computes lanes and displays it on the map editor, so that I can get feedback on it, and then this function will be called upon loading and when unidiffs are applied.

**Comment by nenau-again** [comment]
@bobbens do you have an example of simple code that builds a matrix and performs a few operations on it (product, product with full matrix, linear resolution...) on csparse ? And if possible, it would be perfect to have an example with a multiple right hand side resolution !
It would make me save a lot of time !

**Comment by nenau-again** [comment]
How much ram do you allow me to use ? Because right now, it's 190 Mo, but some operators have to be generated per faction, so it could end up around 300-400 Mo. Is that ok ?
Maybe if I optimize my usage of memory, it could decrease a bit, but no miracle...

**Comment by bobbens** [comment]
> @bobbens do you have an example of simple code that builds a matrix and performs a few operations on it (product, product with full matrix, linear resolution...) on csparse ? And if possible, it would be perfect to have an example with a multiple right hand side resolution !
> It would make me save a lot of time !

Look at [src/economy.c:715](../tree/master/src/economy.c#715) and [src/economy.c:865](../tree/master/src/economy.c#865). Documentation is at [here](https://people.sc.fsu.edu/~jburkardt/c_src/csparse/csparse.html) although it is a bit sparse (I just read the source files included in Naev).

> How much ram do you allow me to use ? Because right now, it's 190 Mo, but some operators have to be generated per faction, so it could end up around 300-400 Mo. Is that ok ?
> Maybe if I optimize my usage of memory, it could decrease a bit, but no miracle...

I don't think 300~400 MiB should be a problem, but we might get complaints on older systems. Does it even use that much with sparse matrices instead of dense ones which I assume you are using right now?

**Comment by ids1024** [comment]
My computer has 16 GB of RAM, so I wouldn't personally mind a few hundred megabytes extra, but that could be a bit more of a problem on lower spec systems.

I don't know exactly how your approach works, but I wonder if this could be done well as a two level process: first, and algorithm that constructs a graph of how systems should be connected by safe lanes, and second an algorithm that generates safe lanes for an individual system, given (among other data) which jump gates the higher level specifies should be connected to lanes.

That would prevent optimal generation in certain ways. But it seems like it could be fairly efficient on CPU and Memory. Then perhaps it could run the higher level algorithm on game load or editing the map, and the lower level one when jumping into a system (though, other methods are possible if it's a bit slow for that).

**Comment by nenau-again** [comment]
Thanks for the use of Csparse. I didn't know it was on master. I guess I'll find a way to get the book as well.

> Does it even use that much with sparse matrices instead of dense ones which I assume you are using right now?

I'm using sparse, but only for matrices that are indeed sparse. There are a few ponderation matrices (per faction unfortunately) that are full. It's for the computation of the topological gradient. I'm writing a document to explain how it works. Maybe I could approach those matrices by sparse ones, I'll see, but only once I get a fully implemented working stuff.

> I wonder if this could be done well as a two level process

A multiscale approach ? That sounds sexy, and is maybe doable. For now however, I will implement the monoscale algorithm that is complicated enough...

**Comment by nenau-again** [comment]
Allright, I think I'm done with the python script. See for example Goddard with multiple factions (empire = green, goddard = blue, dvaered dont produce lanes here):
![goddard_rule_rich](https://user-images.githubusercontent.com/60405677/79050087-f7baba80-7c27-11ea-99a8-5c992ee8cb5d.png)

Or Raelid (I added a rule to prevent factions to build a lane on foreign assets
![raelid](https://user-images.githubusercontent.com/60405677/79050102-0f923e80-7c28-11ea-8c78-6eb37c60dfb1.png)

Memory usage is 200Mo, but run time is about 10s, which is not so small (I have 8 Intel Core i5-8250U CPU @ 1.60GHz). There will be a bit of improvement when skipping to C, but probably no miracle, unless I find some trick somewhere.

I've looked at CSparse, but it misses many things I need on interactions between full and sparse matrices (they only have sparse and I don't want to re-code full matrices stuff), and multi-right hand side stuff. I'll look at LAPACK and ~=UWTFPACK... At least, I have now one interesting book more in my collection about linear algebra.

**Comment by ids1024** [comment]
> Memory usage is 200Mo, but run time is about 10s, which is not so small (I have 8 Intel Core i5-8250U CPU @ 1.60GHz). 

That's not terrible, but it sounds high for running at game load.

>  There will be a bit of improvement when skipping to C, but probably no miracle, unless I find some trick somewhere.

I guess that depends on if there are many loops. Supposing there are, those can (if possible) be replaced with individual numpy operation, or the same code in C will be faster. Otherwise, the same approach in C is probably not really any faster.

I still wonder if it would be possible to do it at a per-system level, after calculating (or manually defining) the graph of how systems are connected. It seems like that's more how a real-world transportation infrastructure project like this would actually be planned, anyway (not that it probably matters).

Continuing to think about it in real world terms, it also seems (though it may not be useful) that factions would build their safe lanes separately. I suppose it might be not be hard to run the algorithm separately for each faction? That should save peak memory usage; though it might take longer. Probably not useful; just a random thought.

**Comment by nenau-again** [comment]
> I guess that depends on if there are many loops.

Each iteration takes about .7s and there are 16 iterations. In each iteration, there is .1s of python loop, that will totally disappear in C, .2s of full matrices multiplications, that will stay the same and .4s  of direct and adjoint sparse inversion. Those .4s are meant to be reduced by 2 means:
* In scipy, there is no easy way to re-use cholesky factorization between direct and adjoint problem, while on any linalg library, it's possible.
* I'm not sure scipy handles multiple right hand side well. If it does not, we'll gain about .2-.3s

> I suppose it might be not be hard to run the algorithm separately for each faction

Actually, my algo consists of 1 optimization algo per faction running sumultaneously, but each of them has a different cost-function to minimize. So I guess one can say that factions build their lanes separately. It's simply that, for performance reasons, they all use the same direct and adjoint problem.

**Comment by nenau-again** [comment]
First the good news:
Thanks to a few tricks, I reduced cpu time to 8s...
Here is a document that explains the approach. I guess/hope people who have a few knowledge in optimization and linear algebra can read it. Don't hesitate if you see errors.
[lanes_generation.pdf](https://github.com/naev/naev/files/4496967/lanes_generation.pdf)

Then the bad ones:
It appears that making sparse matrices dialog with dense matrices in C looks quite complicated if we want it to be efficient. For example, even in scipy, there is some stuff that is not really well handled, and people in scipy are far better than me in that field. I know a few linear-algebra-though-guys, so I will ask them advice, but honestly this does really look complicated.

What is more, the computation time is not really compatible with what we want to do. For these reasons, I tend to be now in favour of the "generate the lanes once and put them in ndata" solution. This is of course less cool than having it being generated online after each unidiff, but actually unidiffs should not drastically change the configuration of the galaxy. The only thing that may change at a big scale is assets ownership, and we could simply keep the lanes as they are, but switch the faction that owns them.

And about the multiscale approach, I did not find a nice way to make it that would be easy to use. Maybe my approach is not the best choice for lanes generation after all, but it's quite easy to use (once implemented) and does fairly nice job in my opinion.

**Comment by ids1024** [comment]
> It appears that making sparse matrices dialog with dense matrices in C looks quite complicated if we want it to be efficient.

**Comment by ids1024** [comment]
> Here is a document that explains the approach. I guess/hope people who have a few knowledge in optimization and linear algebra can read it. Don't hesitate if you see errors.

I'll assume any approach is good if it's documented in an 8 page paper typeset with LaTeX... More seriously, I can probably follow that, but it might take a bit of time.

> It appears that making sparse matrices dialog with dense matrices in C looks quite complicated if we want it to be efficient.

I can see that. It seems like C would be kind of a painful language for linear algebra with any library, and for that reason, there are likely fewer libraries in C.

Can we assume someone building Naev has a Fortran compiler available? :laughing: Or maybe there are some decent C++ libraries.

> For these reasons, I tend to be now in favour of the "generate the lanes once and put them in ndata" solution.

In most regards, this seems perfectly fine. But it does seem like it would make things harder for non-programmers contributing game content.

**Comment by nenau-again** [comment]
> it does seem like it would make things harder for non-programmers contributing game content.

Yes, that's definitely a drawback. But the script could be launched automatically from map editor with a suitable button, I guess. The problem is that it would require contributors to have python3 with scipy installed, and recent experience with the polygon generator have showed that it can sometimes be quite tricky...

**Comment by nenau-again** [comment]
I join my (dirty) python implementation so that anyone can check it out.
The idea for making the code less CPU-costly is to uncouple the different systems (function `optimizeLanes` would only consider the assets and jumps in the current system). I'm pretty sure it would ease computations by a lot, but some of the above-discussed refinements will be lost.

[lanes_generator.py.zip](https://github.com/naev/naev/files/6767617/lanes_generator.py.zip)

**Comment by bobbens** [comment]
Follow up. This is being implement in the `use_lanes` branch and has most things done already. Stuff that is remaining:

- [ ]  Traders don't use safe lanes
- [ ] AI doesn't use safe lanes for their final destination (planet / hyperspace), only when loitering (would probably need a goal or something)
- [ ] Pirates try to use safe lanes also. They should probably avoid them and spawn away from them. Having them stealth and stuff would also be cool.
- [ ] loiter behaviour is very crude when choosing points of interest
- [ ] autonav doesn't use safe lanes (how to let the player choose to use or not?)
- [ ] cargo missions don't consider safe lanes (should probably consider them a bit)

**Comment by UncombedCoconut** [comment]
Some autonav control ideas:

1. In-system, right click for a direct flight; on the map, `[Autonav]` splits into `[Autonav][Mode]` or `[Autonav][…]`, with the button dropdown letting the player put it in safe mode.
2. Checkbox in the Options menu right next to "Stop Speedup At"
3. Hybrid of these: one autonav mode with toggles discoverable on either screen.

Perhaps the tutorial can saw a few words; I just don't want to make it too long or rely on players revisiting it.

In terms of cargo, are we maybe talking about adjusting the "piracy risk" calculation and perhaps adding some text to a rush mission's warning? Overall I like the idea of rush mission offers *not* being concerned with your safety; it can be a feature that the higher reward is only available if you're willing to take a dangerous shortcut.

**Comment by nenau-again** [comment]
I've got an idea that would ease/make possible piracy in factions-controlled space.

It consists in adding cargo ship in unsafe space, that do not actively attack people or anything, but provide various services to everyone willing to pay (ie pirates and the player), and of course remain unattacked by pirates. Their services could be:
* Refuelling, repair, supply services (not free)
* Buy stolen cargo
* Propose some missions
* Maybe sell lesser outfits
* Maybe provide save/load somehow. It would be an alternative to landing, and would maybe require that no enemy ship is in a certain radius (to avoid being a magic option to clear space)

The interactions with the player could be of different types:
* Use their services
* Loot them. But then you'll have problems with the clans
* Chase them to get rewarded by the factions

I believe that a combination of that and jumping in stealth mode could make piracy a viable option (and maybe we'll even be able to spawn the pirate ships fairly (ie not out of nowhere) at jumps, in stealth mode.)

**Comment by bobbens** [comment]
>     1. In-system, right click for a direct flight; on the map, `[Autonav]` splits into `[Autonav][Mode]` or `[Autonav][…]`, with the button dropdown letting the player put it in safe mode.

Something like this could work. Right click, small menu pops so you can set default, if you don't in X seconds it goes away and just uses what was previously set. It's a bit tricky to make it not get in the way while still being usable.

>     2. Checkbox in the Options menu right next to "Stop Speedup At"

I can imagine players wanting to disable / enable it depending on whether or not they have contraband or the likes, so this might not work in that case.

>     3. Hybrid of these: one autonav mode with toggles discoverable on either screen.

This could also work.

> Perhaps the tutorial can saw a few words; I just don't want to make it too long or rely on players revisiting it.

Tutorial should be chopped up into smaller ones that activate conditionally or delayed, and ask the player if they want to reactivate them when they haven't played in a while.

> In terms of cargo, are we maybe talking about adjusting the "piracy risk" calculation and perhaps adding some text to a rush mission's warning? Overall I like the idea of rush mission offers _not_ being concerned with your safety; it can be a feature that the higher reward is only available if you're willing to take a dangerous shortcut.

Well it depends on how important safe lanes become. In the case leaving them is basically a death sentence, this would mean that they should consider it more seriously than if it's just something to be avoidable. It can either be implemented as more lax times, or larger rewards, or a mixture of both.

> It consists in adding cargo ship in unsafe space, that do not actively attack people or anything, but provide various services to everyone willing to pay (ie pirates and the player), and of course remain unattacked by pirates. Their services could be:
>     * Refuelling, repair, supply services (not free)
>     * Buy stolen cargo
>     * Propose some missions
>     * Maybe sell lesser outfits
>     * Maybe provide save/load somehow. It would be an alternative to landing, and would maybe require that no enemy ship is in a certain radius (to avoid being a magic option to clear space)
> 
> The interactions with the player could be of different types:
>     * Use their services
>     * Loot them. But then you'll have problems with the clans
>     * Chase them to get rewarded by the factions
> 
> I believe that a combination of that and jumping in stealth mode could make piracy a viable option (and maybe we'll even be able to spawn the pirate ships fairly (ie not out of nowhere) at jumps, in stealth mode.)

I think that could be interesting. I've toyed with the idea of having small temporary bases playing a role like that, this would just make it a ship. Not sure what would be better, but temporary small bases might be more natural than using ships. Maybe it would be better to do this in a separate "Make Piracy Viable"-type proposal?

**Comment by bobbens** [comment]
With the internal consistency now provided within systems for lanes I think this issue is mainly solved. AI behaviour still needs to be improved, but we don't have the inconsistencies we had a while back. There's still the consideration of making cargo missions and autonav use lanes ,but it doesn't feel as necessary as it did originally. I'll close and we can open new issues on particulars as necessary. Thanks all!
